pkg/share/manager/cs3/cs3.go

```diff
@@ -0,0 +1,558 @@
+type ReceivedShareMetadata struct {
+	State      collaboration.ShareState `json:"state"`
+	MountPoint *provider.Reference      `json:"mountpoint"`
+}
+
+func init() {
+	registry.Register("cs3", NewDefault)
+}
+
+type config struct {
+	GatewayAddr       string `mapstructure:"gateway_addr"`
+	ProviderAddr      string `mapstructure:"provider_addr"`
+	ServiceUserID     string `mapstructure:"service_user_id"`
+	ServiceUserIdp    string `mapstructure:"service_user_idp"`
+	MachineAuthAPIKey string `mapstructure:"machine_auth_apikey"`
+}
+
+// NewDefault returns a new manager instance with default dependencies
+func NewDefault(m map[string]interface{}) (share.Manager, error) {
+	c := &config{}
+	if err := mapstructure.Decode(m, c); err != nil {
+		err = errors.Wrap(err, "error creating a new manager")
+		return nil, err
+	}
+
+	s, err := metadata.NewCS3Storage(c.GatewayAddr, c.ProviderAddr, c.ServiceUserID, c.ServiceUserIdp, c.MachineAuthAPIKey)
+	if err != nil {
+		return nil, err
+	}
+	indexer := indexer.CreateIndexer(s)
+
+	return New(s, indexer)
+}
+
+// New returns a new manager instance
+func New(s Storage, indexer Indexer) (*Manager, error) {
+	return &Manager{
+		storage:     s,
+		indexer:     indexer,
+		initialized: false,
+	}, nil
+}
+
+func (m *Manager) initialize() error {
+	if m.initialized {
+		return nil
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	if m.initialized { // check if initialization happened while grabbing the lock
+		return nil
+	}
+
+	err := m.storage.Init(context.Background(), "cs3-share-manager-metadata")
+	if err != nil {
+		return err
+	}
+	if err := m.storage.MakeDirIfNotExist(context.Background(), "shares"); err != nil {
+		return err
+	}
+	if err := m.storage.MakeDirIfNotExist(context.Background(), "metadata"); err != nil {
+		return err
+	}
+	err = m.indexer.AddIndex(&collaboration.Share{}, option.IndexByFunc{
+		Name: "OwnerId",
+		Func: indexOwnerFunc,
+	}, "Id.OpaqueId", "shares", "non_unique", nil, true)
+	if err != nil {
+		return err
+	}
+	err = m.indexer.AddIndex(&collaboration.Share{}, option.IndexByFunc{
+		Name: "GranteeId",
+		Func: indexGranteeFunc,
+	}, "Id.OpaqueId", "shares", "non_unique", nil, true)
+	if err != nil {
+		return err
+	}
+	m.initialized = true
+	return nil
+}
+
+// Share creates a new share
+func (m *Manager) Share(ctx context.Context, md *provider.ResourceInfo, g *collaboration.ShareGrant) (*collaboration.Share, error) {
+	if err := m.initialize(); err != nil {
+		return nil, err
+	}
+	user := ctxpkg.ContextMustGetUser(ctx)
+	ts := utils.TSNow()
+
+	share := &collaboration.Share{
+		Id: &collaboration.ShareId{
+			OpaqueId: uuid.NewString(),
+		},
+		ResourceId:  md.Id,
+		Permissions: g.Permissions,
+		Grantee:     g.Grantee,
+		Owner:       md.Owner,
+		Creator:     user.Id,
+		Ctime:       ts,
+		Mtime:       ts,
+	}
+
+	data, err := json.Marshal(share)
+	if err != nil {
+		return nil, err
+	}
+
+	fn := path.Join("shares", share.Id.OpaqueId)
```

Comment:
I saw a few places where we hardcode path segments like this. I wonder if we should not have a central place where we define these instead of hardcoding.

Maybe not really part of this PR..

Reply:
@dragotin I just pushed a commit which changes the remaining places to also make use of `shareFilename()` which encapsulates the knowledge about the directoy layout.
Not sure if that's what you meant, but I don't think we need to move this to a central place inside the repository as the layout is only relevant to the share manager itself.